drud/ddev
# #3456 - add project type "node" and configure nginx as reverse proxy for it

This PR aims at providing a solution for running Node.js apps behind a proxy with DDEV. 

See https://github.com/drud/ddev/issues/3456

## The Problem/Issue/Bug:
Currently it is not possible to use ddev to start directly a project unless one configures a custom nginx config with reverse proxy settings.

## How this PR Solves The Problem:
1. A new project type "node" is being added that can be specified in the .ddev/config.yml.
2. The new project type comes without any configuration, but a dedicated nginx config preconfigured as reverse proxy.
3. The reverse proxy will initially forward to https://localhost:3000 so you would have to start your application on that port. This can be changed later, but should be sufficient for an initial implementation.

Further with the current changes it is no longer required to set `disable_settings_management: true` in the config.yml of the project.
## Manual Testing Instructions:
The following config.yml configuration should be enough to run a NextJS app using express:

```
name: my-node-app
type: node
webserver_type: nginx-fpm
nodejs_version: "16"
omit_containers: ['db', 'dba']
hooks:
  post-start:
    - exec: NODE_ENV=development PORT=3000 node server.js
```

If you are using express-ipfilter you might need something like this to allow the request initiator's address after a check that it is running on a dev environment:
```
    if (dev) {
      return IpFilter(allowedIpList, {
        mode: 'allow',
        // Allow the ddev private addresses.
        trustProxy: 'uniquelocal',
      })(req, res, _next);
    }
```
 
## Automated Testing Overview:
No test coverage yet.

## Related Issue Link(s):
https://github.com/drud/ddev/issues/3456

## Release/Deployment notes:
No.



<a href="https://gitpod.io/#https://github.com/drud/ddev/pull/4006"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
